### PR TITLE
feat: odp segment manager

### PR DESCRIPTION
**Pull request by Mat001**
Summary
-------

-  ODP Segment manager for Python-sdk
- Provides an internal api for fetching segments with given apiKey, apiHost, customer key, customer value, segments subset and options.
- See acceptance criteria in the ticket OASIS-8405 below.

Test plan
---------
- added unit tests
- later FSC tests

Issues
------

-  [OASIS-8405 Add OdpSegmentManager](https://optimizely.atlassian.net/browse/OASIS-8405)

### CODE REVIEW

File: tests/test_odp_segment_manager.py
Comment (jaeopt) [inline tests/test_odp_segment_manager.py:49]: Can we also validate that graphQLApi.fetchSegment() is not called in this case?

File: tests/test_odp_segment_manager.py
Comment (jaeopt) [inline tests/test_odp_segment_manager.py:94]: Can we also validate that graphQLApi.fetchSegment() is not called in this case?

**Review by jaeopt — CHANGES_REQUESTED** [review]
A few changes suggested.

File: tests/test_odp_segment_manager.py
Comment (Mat001) [reply tests/test_odp_segment_manager.py:50]: Check added - see line 51 below

File: tests/test_odp_segment_manager.py
Comment (Mat001) [reply tests/test_odp_segment_manager.py:99]: yes, check added, see line 98

File: optimizely/helpers/enums.py
Comment (Mat001) [reply optimizely/helpers/enums.py]: done

File: optimizely/odp/odp_segment_manager.py
Comment (Mat001) [reply optimizely/odp/odp_segment_manager.py]: fixed

File: optimizely/odp/odp_segment_manager.py
Comment (Mat001) [reply optimizely/odp/odp_segment_manager.py]: fixed

File: optimizely/odp/odp_segment_manager.py
Comment (Mat001) [reply optimizely/odp/odp_segment_manager.py]: Fixed, changed into options: list[str]

File: optimizely/odp/odp_segment_manager.py
Comment (Mat001) [reply optimizely/odp/odp_segment_manager.py]: thx, fixed

**Comment by Mat001** [comment]
@jaeopt Fixed your comments. Note that I will need to wait for Andy to merge OdpEventManager and then I will add couple of things to this PR -> refactor some tests. You can review again after I'm done with that.

**Comment by Mat001** [comment]
@andrewleap-optimizely Got your changes in as well. Feel free to give feedback or wait until I make final changes to those tests after the OdpEventManager merge.

**Comment by andrewleap-optimizely** [comment]
> ...wait until I make final changes to a few tests and odp event after the OdpEventManager merge.

The suggested test changes blocked by `OdpEventManager` I was referring to are for `ZaiusRestApiManager` 😃.

**Review by andrewleap-optimizely — CHANGES_REQUESTED** [review]
Looks really good! Just a couple little things

File: tests/test_odp_segment_manager.py
Comment (Mat001) [reply tests/test_odp_segment_manager.py:34]: Good point

**Review by andrewleap-optimizely — CHANGES_REQUESTED** [review]
Just a few more minor suggestions

File: optimizely/odp/odp_options.py
Comment (jaeopt) [inline optimizely/odp/odp_options.py:22]: Can we change the file name to "optimizely_odp_option" to be consistent with "optimizely_decide_option"?

**Review by jaeopt — APPROVED** [review]
LGTM with a nit. Can you remove "WIP" from the PR title?

File: optimizely/odp/odp_segment_manager.py
Comment (Mat001) [reply optimizely/odp/odp_segment_manager.py:75]: @andrewleap-optimizely Not sure what you mean, can you clarify? If we don't hit the cache we miss it no?
U mean if we don't hit the cache there is the third option that is not a cache miss?

Rb segment manager has the same btw

Comment (andrewleap-optimizely) [reply optimizely/odp/odp_segment_manager.py:75]: I mean that the cache might not have been checked, because `ignore_cache` was set.

Yea I should change it in ruby and probably Java where I pulled it from 😀

Comment (andrewleap-optimizely) [reply optimizely/odp/odp_segment_manager.py:75]: Something like this:
```python
if not ignore_cache and not reset_cache:
    segments = self.segments_cache.lookup(cache_key)
    if segments:
        self.logger.debug('ODP cache hit. Returning segments from cache.')
        return segments
    self.logger.debug('ODP cache miss.')

self.logger.debug('Making a call to ODP server.')
```
Otherwise looking at the logs with a disabled cache may confuse the reader into thinking the cache is being used.

Comment (Mat001) [reply optimizely/odp/odp_segment_manager.py:75]: I see, yeah, that makes total sense. good catch

**Review by andrewleap-optimizely — APPROVED** [review]
lgtm

**Pull request merged**
